File: Binary tree/Depth_first_Search/java/iterativeInOrder.java


class IterativeInOrder{
    public List<Integer> inOrderTraversal(Node root){
        List<Integer> result = new ArrayList<>();
        Stack<Node> st = new Stack<Node>();

        Node node = root;

        while(true){
            //if the node is not null we push it to the stack
            if(node !=null){
                stack.push(node);
                //now node becomes left 
                node = node.left;
            }
            else{
                //if the stack is empty so no need to further search so we break 
                if(stack.isEmpty){
                    break;
                }

                node = node.pop()
                result.add(node.val);
                node = node.right
            }

        }
        return result;

    } 
}

File: Binary tree/Depth_first_Search/javascript/IterativeInOrder.js
//inorder traversal using iterative method = left root right

class IterativeInorder {
  iterative(root) {
    let stack = [];
    let result = [];

    let node = root;

    while (true) {
      if (node !== null) {
        stack.push(node);
        node = node.left;
      } else {
        if (stack.length === 0) {
          break;
        }
        node = stack.pop();
        result.push(node.val);
        node = node.right;
      }
    }
    return result;
  }
}
